addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py

```diff
@@ -31,3 +31,77 @@ def test_mailing_schedule_date(self):
         self.assertEqual(mailing.schedule_date, datetime(2021, 4, 30, 9, 0))
         self.assertEqual(mailing.schedule_type, 'scheduled')
         self.assertEqual(mailing.state, 'in_queue')
+
+    def test_mailing_next_departure(self):
+        # test if mailing.mailing.next_departure is correctly set taking into account
+        # presence of implicitly created cron triggers (since odoo v15). These should
+        # launch cron job before its schedule nextcall datetime (if scheduled_date < nextcall)
+
+        # case were user click on "Send" button (action_launch)
+        start_time = datetime.now()
+        mailing = self.env['mailing.mailing'].create({
+            'name': 'mailing',
+            'subject': 'some subject',
+            'mailing_model_id': self.env['ir.model']._get('res.partner').id,
+            'state' : 'draft'
+        })
+
+        cron_job = self.env.ref('mass_mailing.ir_cron_mass_mailing_queue').sudo()
+        cron_job_nextcall = datetime(2099, 4, 30, 9, 0) # way in the future
+        cron_job.write({'nextcall' : cron_job_nextcall})
+
+        mailing.action_launch()
+        end_time = datetime.now() + relativedelta(seconds=30)
+        test_now_schedule_date = True if start_time <= mailing.next_departure <= end_time else False
+        # assert that the schedule_date and schedule_type fields are correct and that the mailing is put in queue
+        self.assertEqual(test_now_schedule_date, True)
+        self.assertIsNot(mailing.schedule_date, cron_job_nextcall)
+        self.assertEqual(mailing.schedule_type, 'now')
+        self.assertEqual(mailing.state, 'in_queue')
+
+        # test cancel button
+        mailing.action_cancel()
+
+        # assert if cancel reset fields as expected
+        self.assertEqual(mailing.state, 'draft')
+        self.assertEqual(mailing.schedule_type, 'now')
+        self.assertEqual(mailing.schedule_date, False)
+        self.assertEqual(mailing.next_departure, False)
+
+        # case where client uses schedule wizard to chose a date between now and cron.job nextcall
+        # create a schedule date wizard
+        wizard_form = Form(
+            self.env['mailing.mailing.schedule.date'].with_context(default_mass_mailing_id=mailing.id))
+
+        # set a schedule date in near future
+        schedule_date_user = datetime.now() + relativedelta(minutes=50)
+        wizard_form.schedule_date = schedule_date_user
+        wizard = wizard_form.save()
+        wizard.action_schedule_date()
```

Comment:
I would just create a new mailing with the proper state. This code seems irrelevant to what is being tested imo.